### ISSUE TITLE
Highlighting range flattening panics

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#4013]
```
thread '<unnamed>' panicked at 'assertion failed: res.iter().zip(res.iter().skip(1)).all(|(left, right)|
                                           left.range.end() <=
                                               right.range.start())', crates/ra_ide/src/syntax_highlighting.rs:179:5
```

I get this at in the stdlib file which defines `thread_local` (just goto definition on it).

cc @ltentrup

**Comment by ltentrup** [comment]
This assertion checks that the flattened ranges are non-overlapping. My guess is that there are overlapping ranges already in the AST:  in the flattening logic there is an assertion that the child ranges are sub ranges of the parent range.

**Comment by ltentrup** [comment]
Here is a reduced example:

```rust
#[stable(feature = "rust1")]
macro_rules! thread_local {
```

It seems that first the `macro_rules!` is added and later on the attribute, where the ranges of the attribute are before the ranges for `macro_rules!`. A simple fix for this specific problem would be to sort the ranges afterward.

**Comment by ltentrup** [comment]
This can be closed as it was fixed in #4022
